File: backend/app/routes/stats.py
from collections import Counter

from fastapi import APIRouter, Depends
from sqlalchemy import select, func
from sqlalchemy.orm import Session

from ..database import get_db
from ..models import Card

router = APIRouter(prefix="/api/stats", tags=["stats"])


@router.get("")
def get_stats(db: Session = Depends(get_db)):
    """Compute collection statistics."""
    cards = db.scalars(select(Card)).all()

    if not cards:
        return {"empty": True}

    total_unique = len(cards)
    total_copies = sum(c.quantity for c in cards)
    distinct_card_ids = len({c.card_id for c in cards})

    # --- Price helpers ---
    def best_price(c):
        for p in [c.price_cm_median, c.price_cm_avg, c.price_cm_min, c.price_cardmarket]:
            if p is not None:
                return p
        return 0.0

    total_value = sum(best_price(c) * c.quantity for c in cards)

    priced_cards = [c for c in cards if best_price(c) > 0]
    avg_card_value = total_value / total_copies if total_copies else 0

    # --- Distributions (weighted by quantity) ---
    rarity_dist = Counter()
    condition_dist = Counter()
    lang_dist = Counter()
    type_dist = Counter()
    attribute_dist = Counter()
    race_dist = Counter()
    level_dist = Counter()
    archetype_dist = Counter()
    set_dist = Counter()
    frame_type_dist = Counter()

    for c in cards:
        q = c.quantity
        rarity_dist[c.rarity] += q
        condition_dist[c.condition] += q
        lang_dist[c.lang] += q
        type_dist[c.type] += q
        frame_type_dist[c.frame_type] += q
        if c.attribute:
            attribute_dist[c.attribute] += q
        if c.race:
            race_dist[c.race] += q
        if c.level is not None:
            level_dist[str(c.level)] += q
        if c.archetype:
            archetype_dist[c.archetype] += q
        if c.set_code:
            prefix = c.set_code.split("-")[0]
            set_dist[prefix] += q

    # --- Top cards by value ---
    top_valuable = sorted(cards, key=lambda c: best_price(c), reverse=True)[:10]
    top_valuable_data = [
        {
            "name": c.name,
            "set_code": c.set_code,
            "rarity": c.rarity,
            "price": round(best_price(c), 2),
            "quantity": c.quantity,
            "image_url": c.image_url,
            "total_value": round(best_price(c) * c.quantity, 2),
        }
        for c in top_valuable if best_price(c) > 0
    ]

    # --- Value by rarity ---
    value_by_rarity = Counter()
    for c in cards:
        value_by_rarity[c.rarity] += best_price(c) * c.quantity
    value_by_rarity = {k: round(v, 2) for k, v in value_by_rarity.most_common(15)}

    # --- Value by language ---
    value_by_lang = Counter()
    for c in cards:
        value_by_lang[c.lang] += best_price(c) * c.quantity
    value_by_lang = {k: round(v, 2) for k, v in value_by_lang.most_common()}

    # --- ATK/DEF scatter data (for monsters) ---
    atk_def_data = []
    seen_atk_def = set()
    for c in cards:
        if c.atk is not None and c.def_ is not None:
            key = (c.card_id, c.atk, c.def_)
            if key not in seen_atk_def:
                seen_atk_def.add(key)
                atk_def_data.append({
                    "name": c.name,
                    "atk": c.atk,
                    "def": c.def_,
                    "level": c.level,
                })

    # --- Price distribution histogram ---
    price_ranges = {"0": 0, "0.01-0.50": 0, "0.51-1": 0, "1-5": 0, "5-10": 0, "10-25": 0, "25-50": 0, "50-100": 0, "100+": 0}
    for c in cards:
        p = best_price(c)
        q = c.quantity
        if p == 0:
            price_ranges["0"] += q
        elif p <= 0.50:
            price_ranges["0.01-0.50"] += q
        elif p <= 1:
            price_ranges["0.51-1"] += q
        elif p <= 5:
            price_ranges["1-5"] += q
        elif p <= 10:
            price_ranges["5-10"] += q
        elif p <= 25:
            price_ranges["10-25"] += q
        elif p <= 50:
            price_ranges["25-50"] += q
        elif p <= 100:
            price_ranges["50-100"] += q
        else:
            price_ranges["100+"] += q

    # --- Collection growth (cards added over time) ---
    cards_with_date = sorted(
        [(c.created_at.strftime("%Y-%m") if c.created_at else None, c.quantity) for c in cards],
        key=lambda x: x[0] or ""
    )
    growth_by_month = Counter()
    for month, qty in cards_with_date:
        if month:
            growth_by_month[month] += qty

    return {
        "empty": False,
        "overview": {
            "total_unique": total_unique,
            "total_copies": total_copies,
            "distinct_cards": distinct_card_ids,
            "total_value": round(total_value, 2),
            "avg_card_value": round(avg_card_value, 2),
            "priced_count": len(priced_cards),
            "unpriced_count": total_unique - len(priced_cards),
        },
        "distributions": {
            "rarity": dict(rarity_dist.most_common()),
            "condition": dict(condition_dist.most_common()),
            "language": dict(lang_dist.most_common()),
            "type": dict(type_dist.most_common()),
            "attribute": dict(attribute_dist.most_common()),
            "race": dict(race_dist.most_common(20)),
            "level": dict(sorted(level_dist.items(), key=lambda x: int(x[0]))),
            "archetype": dict(archetype_dist.most_common(20)),
            "set": dict(set_dist.most_common(20)),
            "frame_type": dict(frame_type_dist.most_common()),
        },
        "top_valuable": top_valuable_data,
        "value_by_rarity": value_by_rarity,
        "value_by_lang": value_by_lang,
        "price_distribution": price_ranges,
        "atk_def_scatter": atk_def_data,
        "growth_by_month": dict(sorted(growth_by_month.items())),
    }

File: frontend/js/stats.js
// =============================================
// Statistics view — charts and KPIs
// =============================================

import { API, cardImgUrl } from "./shared.js";

let _statsCharts = {};

function _destroyCharts() {
    Object.values(_statsCharts).forEach(c => c.destroy());
    _statsCharts = {};
}

const CHART_COLORS = [
    "#7c5cfc", "#22c55e", "#ef4444", "#3b82f6", "#eab308", "#f97316",
    "#06b6d4", "#ec4899", "#8b5cf6", "#14b8a6", "#f43f5e", "#a855f7",
    "#0ea5e9", "#84cc16", "#d946ef", "#fbbf24", "#6366f1", "#10b981",
    "#e11d48", "#059669",
];

const ATTR_COLORS = {
    DARK: "#9333ea", LIGHT: "#fbbf24", FIRE: "#ef4444", WATER: "#3b82f6",
    EARTH: "#a16207", WIND: "#22c55e", DIVINE: "#f59e0b",
};

const COND_COLORS = {
    "Mint": "#00bcd4", "Near Mint": "#4caf50", "Excellent": "#8bc34a",
    "Good": "#fdd835", "Light Played": "#ff9800", "Played": "#ff5722", "Poor": "#f44336",
};

function _chartDefaults() {
    return {
        responsive: true,
        maintainAspectRatio: false,
        plugins: {
            legend: {
                labels: { color: "#9ca3af", font: { family: "'Inter', sans-serif", size: 11 } },
            },
        },
    };
}

function _doughnut(canvasId, labels, data, colors) {
    const ctx = document.getElementById(canvasId);
    _statsCharts[canvasId] = new Chart(ctx, {
        type: "doughnut",
        data: {
            labels,
            datasets: [{
                data,
                backgroundColor: colors || labels.map((_, i) => CHART_COLORS[i % CHART_COLORS.length]),
                borderColor: "rgba(0,0,0,0.3)",
                borderWidth: 1,
            }],
        },
        options: {
            ..._chartDefaults(),
            cutout: "55%",
            plugins: {
                ..._chartDefaults().plugins,
                legend: {
                    position: "right",
                    labels: { color: "#9ca3af", font: { family: "'Inter', sans-serif", size: 11 }, padding: 8 },
                },
            },
        },
    });
}

function _bar(canvasId, labels, data, color, label) {
    const ctx = document.getElementById(canvasId);
    _statsCharts[canvasId] = new Chart(ctx, {
        type: "bar",
        data: {
            labels,
            datasets: [{
                label: label || "",
                data,
                backgroundColor: typeof color === "string" ? color : labels.map((_, i) => (color || CHART_COLORS)[i % (color || CHART_COLORS).length]),
                borderRadius: 4,
            }],
        },
        options: {
            ..._chartDefaults(),
            indexAxis: "x",
            plugins: { ..._chartDefaults().plugins, legend: { display: false } },
            scales: {
                x: { ticks: { color: "#6b7280", font: { size: 10 } }, grid: { color: "rgba(255,255,255,0.04)" } },
                y: { ticks: { color: "#6b7280" }, grid: { color: "rgba(255,255,255,0.04)" } },
            },
        },
    });
}

function _horizontalBar(canvasId, labels, data, colors) {
    const ctx = document.getElementById(canvasId);
    const barHeight = 28;
    const minHeight = 200;
    const neededHeight = Math.max(minHeight, labels.length * barHeight + 40);
    const wrapper = ctx.parentElement;
    wrapper.style.height = neededHeight + "px";

    _statsCharts[canvasId] = new Chart(ctx, {
        type: "bar",
        data: {
            labels,
            datasets: [{
                data,
                backgroundColor: colors || labels.map((_, i) => CHART_COLORS[i % CHART_COLORS.length]),
                borderRadius: 4,
            }],
        },
        options: {
            ..._chartDefaults(),
            indexAxis: "y",
            plugins: { ..._chartDefaults().plugins, legend: { display: false } },
            scales: {
                x: { ticks: { color: "#6b7280" }, grid: { color: "rgba(255,255,255,0.04)" } },
                y: { ticks: { color: "#9ca3af", font: { size: 11 }, autoSkip: false }, grid: { display: false } },
            },
        },
    });
}

export async function loadStats() {
    const loading = document.getElementById("stats-loading");
    const empty = document.getElementById("stats-empty");
    const content = document.getElementById("stats-content");
    loading.hidden = false;
    empty.hidden = true;
    content.hidden = true;

    try {
        const resp = await fetch(`${API}/api/stats`);
        const stats = await resp.json();
        loading.hidden = true;

        if (stats.empty) { empty.hidden = false; return; }

        content.hidden = false;
        _destroyCharts();

        const ov = stats.overview;
        document.getElementById("kpi-total-value").textContent = `${ov.total_value.toFixed(2)}\u20AC`;
        document.getElementById("kpi-unique").textContent = ov.total_unique.toLocaleString();
        document.getElementById("kpi-copies").textContent = ov.total_copies.toLocaleString();
        document.getElementById("kpi-distinct").textContent = ov.distinct_cards.toLocaleString();
        document.getElementById("kpi-avg-value").textContent = `${ov.avg_card_value.toFixed(2)}\u20AC`;
        document.getElementById("kpi-priced").textContent = `${ov.priced_count} / ${ov.unpriced_count}`;

        // Top valuable
        const topList = document.getElementById("top-valuable-list");
        topList.innerHTML = stats.top_valuable.map((c, i) => `
            <div class="top-valuable-item">
                <span class="tv-rank">#${i + 1}</span>
                <img class="tv-img" src="${cardImgUrl(c.image_url)}" alt="${c.name}" loading="lazy">
                <div class="tv-info">
                    <div class="tv-name">${c.name}</div>
                    <div class="tv-meta">${c.set_code || ""} &middot; ${c.rarity}</div>
                </div>
                <div class="tv-price-col">
                    <div class="tv-price">${c.price.toFixed(2)}\u20AC</div>
                    ${c.quantity > 1 ? `<div class="tv-qty">x${c.quantity} = ${c.total_value.toFixed(2)}\u20AC</div>` : ""}
                </div>
            </div>
        `).join("");

        // Doughnut charts
        const dist = stats.distributions;
        _doughnut("chart-rarity", Object.keys(dist.rarity), Object.values(dist.rarity));
        _doughnut("chart-type", Object.keys(dist.type), Object.values(dist.type));
        _doughnut("chart-condition",
            Object.keys(dist.condition), Object.values(dist.condition),
            Object.keys(dist.condition).map(k => COND_COLORS[k] || CHART_COLORS[0]));
        _doughnut("chart-lang", Object.keys(dist.language), Object.values(dist.language));
        _doughnut("chart-attribute",
            Object.keys(dist.attribute), Object.values(dist.attribute),
            Object.keys(dist.attribute).map(k => ATTR_COLORS[k] || CHART_COLORS[0]));
        _doughnut("chart-race", Object.keys(dist.race), Object.values(dist.race));

        // Bar charts
        const rarityLabels = Object.keys(stats.value_by_rarity).map(k => `${k} (${stats.count_by_rarity[k] || 0})`);
        _horizontalBar("chart-value-rarity", rarityLabels, Object.values(stats.value_by_rarity));
        const avgRarityLabels = Object.keys(stats.avg_by_rarity).map(k => `${k} (${stats.count_by_rarity[k] || 0})`);
        _horizontalBar("chart-avg-rarity", avgRarityLabels, Object.values(stats.avg_by_rarity));
        _horizontalBar("chart-value-lang", Object.keys(stats.value_by_lang), Object.values(stats.value_by_lang));
        _bar("chart-price-dist", Object.keys(stats.price_distribution), Object.values(stats.price_distribution), "#7c5cfc");
        _bar("chart-level", Object.keys(dist.level).map(l => `Lv ${l}`), Object.values(dist.level), "#3b82f6");
        _horizontalBar("chart-sets", Object.keys(dist.set), Object.values(dist.set));
        _horizontalBar("chart-archetypes", Object.keys(dist.archetype), Object.values(dist.archetype));

        // ATK/DEF scatter
        const scatterCtx = document.getElementById("chart-atk-def");
        _statsCharts["chart-atk-def"] = new Chart(scatterCtx, {
            type: "scatter",
            data: {
                datasets: [{
                    label: "Monsters",
                    data: stats.atk_def_scatter.map(m => ({ x: m.atk, y: m.def, name: m.name })),
                    backgroundColor: "rgba(124, 92, 252, 0.5)",
                    borderColor: "#7c5cfc",
                    pointRadius: 3,
                    pointHoverRadius: 6,
                }],
            },
            options: {
                ..._chartDefaults(),
                plugins: {
                    ..._chartDefaults().plugins,
                    legend: { display: false },
                    tooltip: { callbacks: { label: ctx => `${ctx.raw.name}: ATK ${ctx.raw.x} / DEF ${ctx.raw.y}` } },
                },
                scales: {
                    x: { title: { display: true, text: "ATK", color: "#6b7280" }, ticks: { color: "#6b7280" }, grid: { color: "rgba(255,255,255,0.04)" } },
                    y: { title: { display: true, text: "DEF", color: "#6b7280" }, ticks: { color: "#6b7280" }, grid: { color: "rgba(255,255,255,0.04)" } },
                },
            },
        });

        // Growth
        const months = Object.keys(stats.growth_by_month);
        const growthData = Object.values(stats.growth_by_month);
        const cumulative = [];
        growthData.reduce((acc, val, i) => { cumulative[i] = acc + val; return cumulative[i]; }, 0);

        const growthCtx = document.getElementById("chart-growth");
        _statsCharts["chart-growth"] = new Chart(growthCtx, {
            type: "line",
            data: {
                labels: months,
                datasets: [
                    { label: "Cumulative", data: cumulative, borderColor: "#7c5cfc", backgroundColor: "rgba(124, 92, 252, 0.1)", fill: true, tension: 0.3, pointRadius: 4, pointBackgroundColor: "#7c5cfc" },
                    { label: "Added", data: growthData, borderColor: "#22c55e", backgroundColor: "rgba(34, 197, 94, 0.1)", fill: true, tension: 0.3, pointRadius: 4, pointBackgroundColor: "#22c55e" },
                ],
            },
            options: {
                ..._chartDefaults(),
                plugins: { ..._chartDefaults().plugins, legend: { labels: { color: "#9ca3af" } } },
                scales: {
                    x: { ticks: { color: "#6b7280" }, grid: { color: "rgba(255,255,255,0.04)" } },
                    y: { ticks: { color: "#6b7280" }, grid: { color: "rgba(255,255,255,0.04)" } },
                },
            },
        });

    } catch (e) {
        console.error("Failed to load stats:", e);
        loading.hidden = true;
        empty.hidden = false;
    }
}

// --- Tabbed chart cards ---
document.querySelectorAll(".stats-tab-bar").forEach(bar => {
    bar.querySelectorAll(".stats-tab").forEach(tab => {
        tab.addEventListener("click", () => {
            const card = bar.closest(".stats-tabbed-card");
            card.querySelectorAll(".stats-tab").forEach(t => t.classList.remove("active"));
            card.querySelectorAll(".stats-tab-panel").forEach(p => p.classList.remove("active"));
            tab.classList.add("active");
            card.querySelector(`#${tab.dataset.tab}`).classList.add("active");
        });
    });
});
